cipher/block-cipher/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+pub use crate::{
+    cipher::DEBlockCipher,
+    suite::{
+        Aes128, Aes128Circuit, Aes128ShareCircuit, BlockCipherCircuit, BlockCipherCircuitSuite,
+        BlockCipherShareCircuit,
+    },
+};
+pub use config::{
+    BlockCipherConfig, BlockCipherConfigBuilder, BlockCipherConfigBuilderError, Role,
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum BlockCipherError {
+    #[error("MuxerError: {0}")]
+    MuxerError(#[from] utils_aio::mux::MuxerError),
+    #[error("GCFactoryError: {0}")]
+    GCFactoryError(#[from] mpc_aio::protocol::garble::factory::GCFactoryError),
+    #[error("GCError: {0}")]
+    GCError(#[from] GCError),
+    #[error("Cipher key labels not set")]
+    KeysNotSet,
+    #[error("Encoder not set")]
+    EncoderNotSet,
+    #[error("Input does not match block length: expected {0}, got {1}")]
+    InvalidInputLength(usize, usize),
+}
+
+#[derive(Clone)]
+pub struct BlockCipherLabels {
+    key_full: FullLabels,
+    key_active: ActiveLabels,
+}
+
+#[async_trait]
+pub trait BlockCipher<Cipher>
+where
+    Cipher: BlockCipherCircuitSuite,
+    Self: Sized + Send,
+{
+    /// Sets the key input labels for the block cipher.
+    ///
+    /// * `labels`: The labels to use for the key input.
+    fn set_keys(&mut self, labels: BlockCipherLabels);
+
+    /// Sets the encoder used to generate the input labels
+    /// used during 2PC.
+    ///
+    /// * `encoder`: The encoder to use
+    fn set_encoder(&mut self, encoder: Arc<Mutex<ChaChaEncoder>>);
+
+    /// Encrypts the given plaintext.
+    ///
+    /// Returns the ciphertext
+    ///
+    /// * `plaintext` - The plaintext to encrypt
+    async fn encrypt_private(&mut self, plaintext: Vec<u8>) -> Result<Vec<u8>, BlockCipherError>;
+
+    /// Encrypts a plaintext provided by the other party
+    ///
+    /// Returns the ciphertext
+    async fn encrypt_blind(&mut self) -> Result<Vec<u8>, BlockCipherError>;
+
+    /// Encrypts a plaintext provided by both parties. Fails if the
+    /// plaintext provided by both parties does not match.
+    ///
+    /// Returns the additive share of the ciphertext
+    ///
+    /// * `plaintext` - The plaintext to encrypt
+    /// * `mask` - The additive share of the mask to use
+    async fn encrypt_share(
+        &mut self,
+        plaintext: Vec<u8>,
+        mask: Vec<u8>,
+    ) -> Result<Vec<u8>, BlockCipherError>;
+}
+
+#[cfg(feature = "mock")]
+pub mod mock {
+    use mpc_aio::protocol::garble::{
+        exec::dual::mock::{MockDualExFollower, MockDualExLeader},
+        factory::dual::mock::{create_mock_dualex_factory, MockDualExFactory},
+    };
+    use mpc_circuits::{BitOrder, Value};
+    use mpc_core::garble::Encoder;
+
+    use super::*;
+
+    pub type MockDEBlockCipherLeader<C> = DEBlockCipher<C, MockDualExFactory, MockDualExLeader>;
+    pub type MockDEBlockCipherFollower<C> = DEBlockCipher<C, MockDualExFactory, MockDualExFollower>;
+
+    pub fn create_labels<C: BlockCipherCircuitSuite>(
```

Comment:
I would expect to see here how the output labels from one GC execution become the key labels for block cipher like it will be in TLSNotary. Is there anything special that the encoder have to be aware of? This will help me see the big picture.